### ISSUE TITLE
Improve wording for the addition of new server

### ISSUE THREAD

**Issue by Porkepix** [vector-im/riot-web#13661]
### Description

The wording for the addition of a new server is confusing. It asks for a server name (Like, how will it appear in the list) while it actually wait the server host/address.

### Steps to reproduce

Try to add a new server

![Screenshot from 2020-05-13 21-10-40](https://user-images.githubusercontent.com/709500/81854673-93f51b80-955e-11ea-893c-719e0dab2764.png)

### Version information

- **Platform**: web (in-browser) or desktop? Both.
- Latest version for both Web and desktop (1.6).

**Comment by t3chguy** [comment]
No. It does take a server name. Like Mozilla.org instead of the resolvable address Mozilla.modular.im

It takes the value of the synapse Server_name parameter

**Comment by Porkepix** [comment]
Okay so things are not completely clear to me. I tried to add different servers to give it a try, and was unable to deduce what should I add here with many servers. Looks like for this one with Mozilla, I got it green out of pure luck.
How is one supposed to get the value that should be put in this field? If wording is correct as you suggest (though I don't see why yet), maybe would it need some more text to hint the user about it?

**Comment by t3chguy** [comment]
It is the second half of a user's Matrix ID, e.g @user:matrix.org (not the address matrix-client.matrix.org), @user:mozilla.org (not the address mozilla.modular.im) etc

**Comment by Porkepix** [comment]
So that's exactly what I meant when I opened the issue. For me that's the host server, and not the name. Whether it is some kind of alias, a loadbalancer endpoint or the real host address does not matter to me, but that's not what I'd call "a name" when it involves a valid domain name.

I would expect a name to be in the form of, well, a name: Mozilla, or whatever organization or nickname name without any form of domain name.

**Comment by t3chguy** [comment]
Right but you can't connect to `mozilla.org` to expect its Matrix server so it isn't an address. 
If you enter mozilla.org into the Custom Homeserver URL field it won't work.

**Comment by Porkepix** [comment]
> Right but you can't connect to `mozilla.org` to expect its Matrix server so it isn't an address.
> If you enter mozilla.org into the Custom Homeserver URL field it won't work.
> 
> Its just consistent wording with the rest of the ecosystem

So, if that's indeed just a display name and not any kind of server, in case of a new Matrix instance where no one have yet established any contact with another Matrix instance, how is the client supposed to find the address of contact?

**Comment by t3chguy** [comment]
Well given how federation works, where servers communicate on an as needed basis and no one server knows all others, this is exactly matched. For federation you need the server_name, which then gets delegation rules applied to it and is the only way to resolve a server for federation.

**Comment by Porkepix** [comment]
I don't get the explanation. Let's say I pick a random, small and probably not much used server, like matrix.fdn.fr I found out. It looks like adding this doesn't work:

![Screenshot from 2020-05-14 13-14-56](https://user-images.githubusercontent.com/709500/81928055-f2fd7380-95e4-11ea-9e67-1f1ca7b9406a.png)

That leads users into a position where they doesn't know what they have to add to browse instance's channel list.

(I was already pointed to this feature as I first just used the search. I used to be a heavy user of Mozilla's irc. I'm not as active anymore but had a one-time need to check something, and it made me notice that a simple search through Riot's search without further configuration only found out less than a dozen of chans; I was expecting for search to propagate to known servers, but I was told the way to do so was to add server in this search section).

**Comment by t3chguy** [comment]
Well the server there actually is `fdn.fr` probably, matrix.fdn.fr is just a routing detail.

Just like you can host a website project.xyz at foobar.com, foobar.org, foobar.xyz, the website's canonical name is project.xyz

**Comment by Porkepix** [comment]
> Well the server there actually is `fdn.fr` probably, matrix.fdn.fr is just a routing detail and may not be the only way to connect to that server _AS A CLIENT_. For federation only the name works.

Except `fdn.fr` doesn't works either…
I probably miss the point completely here for me to still don't understand how a user should know what have to be input here. Maybe some doc could shed some light on it, but I guess my point of clarification still holds even if it's in a form of hints/examples considering, as a technical people, I can't get this (but maybe that due to the technical background, I want too much to understand how does it works).

**Comment by t3chguy** [comment]
The issue with `fdn.fr` is their room directory is not public over federation, it is a setting in Synapse: https://matrix.fdn.fr/_matrix/client/r0/publicRooms

The user should never be exposed to `matrix.fdn.fr` in an ideal world, it is just a connection detail of which many CNAME DNS can point to the same server.

**Comment by Porkepix** [comment]
Which means, it's not supported to have several instances of a Matrix server behind a unique domain name?

That's still unclear from me when, say, a random user want to search through a matrix instance and so add it in this section of Riot, user likely have one information: address of the server (the part after the colon, I guess?).

From that single piece of information, how can they deduce what should be the input of the "Add a new server" field?

**Comment by t3chguy** [comment]
They can't, if it was possible then Riot would do it for you. The only bit of information you can use for federation is the server_name `fdn.fr`

You shouldn't be sharing the address `matrix.fdn.fr` except to power users, all other users should just log in with their username `@user:fdn.fr` which would automatically look up (one way lookup) matrix.fdn.fr and connect there transparently.

**Comment by Porkepix** [comment]
So, you can't for example have `synapse.mydomain.com` in the same time as `dendrite.mydomain.com` because they would conflict over the use of `mydomain.com`, am I right?

To try to sum up if things could be improved on that way or just if the user needs better knowledge to use this:

- Am I correct to assume this field always expect the part that come after the colon in a Matrix handle? Could this be made more clear to the user?
- If server doesn't have a public directory, is it possible to distinguish this error from the one where the server just doesn't exist (or blacklisted your instance, ie. resulting in a timeout)? This means having two different error messages besides this field depending on the case, to hint the user to what's wrong.
- Maybe this could be a useful addition to have some kind of automatic room directory addition. Let me explain myself here. As a technical person used a bit to how things work on the Internet, when I wanted to see Mozilla's available rooms, knowing that the server part of their handles was `:mozilla.org`, my first reaction was simply to go over the Explore rooms button, and to search for `:mozilla.org`. It only found less than a dozen of channels. From a UX standpoint, I think the interface could recognize the server part of a Matrix handle (starting with a colon), check if there's a directory behind this, and browse it/add it automatically. What do you think about this?

**Comment by t3chguy** [comment]
It depends which one of those `mydomain.com` points to using SRV/.well-known


`synapse.mydomain.com` could be for the `otherproject.xyz` Matrix server
The domain/IP that Synapse/Dendrite/whatever runs at doesn't have to be related to the Server Name. The Server Name, e.g `mydomain.com` has to delegate (point) to the server it wants to represent it. That pointer is used for federation.

**Comment by t3chguy** [comment]
> Am I correct to assume this field always expect the part that come after the colon in a Matrix handle?

Yes, this is what I said earlier: https://github.com/vector-im/riot-web/issues/13661#issuecomment-628496193

> If server doesn't have a public directory, is it possible to distinguish this error from the one where the server just doesn't exist (or blacklisted your instance, ie. resulting in a timeout)? This means having two different error messages besides this field depending on the case, to hint the user to what's wrong.

Nope, Synapse just tells us it failed, not anything about how it failed.

> Maybe this could be a useful addition to have some kind of automatic room directory addition. Let me explain myself here. As a technical person used a bit to how things work on the Internet, when I wanted to see Mozilla's available rooms, knowing that the server part of their handles was :mozilla.org, my first reaction was simply to go over the Explore rooms button, and to search for :mozilla.org. It only found less than a dozen of channels. From a UX standpoint, I think the interface could recognize the server part of a Matrix handle (starting with a colon), check if there's a directory behind this, and browse it/add it automatically. What do you think about this?

Feel free to open a new issue to track this.

**Comment by Porkepix** [comment]
> It depends which one of those `mydomain.com` points to using SRV/.well-known
> 
> `synapse.mydomain.com` could be for the `otherproject.xyz` Matrix server
> The domain/IP that Synapse/Dendrite/whatever runs at doesn't have to be related to the Server Name. The Server Name, e.g `mydomain.com` has to delegate (point) to the server it wants to represent it. That pointer is used for federation.

Then how does it manage possible conflicts over names. If say, two different instances try to claim use of the same name?

> > Am I correct to assume this field always expect the part that come after the colon in a Matrix handle?
> 
> Yes, this is what I said earlier: [#13661 (comment)](https://github.com/vector-im/riot-web/issues/13661#issuecomment-628496193)

Okay, so this could see the addition of a hint, then :)


> > If server doesn't have a public directory, is it possible to distinguish this error from the one where the server just doesn't exist (or blacklisted your instance, ie. resulting in a timeout)? This means having two different error messages besides this field depending on the case, to hint the user to what's wrong.
> 
> Nope, Synapse just tells us it failed, not anything about how it failed.

Which imply Synapse answered, hence that there's a server out there. So I'd say it's fairly easy from here to know if a server exist, or if the server have nothing to answer on directory listing tries?

> Feel free to open a new issue to track this.

Okay, I will try to do so, as that's right a lot of discussions occurred here :)

**Comment by t3chguy** [comment]
> Then how does it manage possible conflicts over names. If say, two different instances try to claim use of the same name?

Because as I said, its the `mozilla.org` which points to which server it wants to represent it, the server name which a server has configured it just so it knows what to put it its identifiers, e.g @foobar:mozilla.org

Its a one way pointer like DNS. From the canonical Server Name, e.g `mydomain.com` to `someserver.somewhere.else`

> Which imply Synapse answered, hence that there's a server out there. So I'd say it's fairly easy from here to know if a server exist, or if the server have nothing to answer on directory listing tries?

No, we have to ask our own server to get the result over federation. So our Synapse will *ALWAYS* answer after it is done trying on our behalf.

**Comment by Porkepix** [comment]
> > Feel free to open a new issue to track this.
> 
> Okay, I will try to do so, as that's right a lot of discussions occurred here :)

Opened there: #13670 :)

**Comment by Porkepix** [comment]
> > Then how does it manage possible conflicts over names. If say, two different instances try to claim use of the same name?
> 
> Because as I said, its the `mozilla.org` which points to which server it wants to represent it, the server name which a server has configured it just so it knows what to put it its identifiers, e.g @foobar:mozilla.org
> 
> Its a one way pointer like DNS. From the canonical Server Name, e.g `mydomain.com` to `someserver.somewhere.else`

Okay, so a DNS record.
Thanks for all the explanations!

> > Which imply Synapse answered, hence that there's a server out there. So I'd say it's fairly easy from here to know if a server exist, or if the server have nothing to answer on directory listing tries?
> 
> No, we have to ask our own server to get the result over federation. So our Synapse will _ALWAYS_ answer after it is done trying on our behalf.
> See https://matrix.org/docs/spec/client_server/r0.6.0#get-matrix-client-r0-publicrooms `server` parameter

And could we imagine the server could answer to the request more precisely? Is so, is this an issue I can/should open? In some kind of spec repo? Or directly into Synapse, Dendrite, whatever server repos?

**Comment by t3chguy** [comment]
> And could we imagine the server could answer to the request more precisely? Is so, is this an issue I can/should open? In some kind of spec repo? Or directly into Synapse, Dendrite, whatever server repos?

Yes for sure it'd be nice if our server told us why the request failed in some detail, the spec repo is https://github.com/matrix-org/matrix-doc/issues
